### PR TITLE
Navigator with tests and demo

### PR DESCRIPTION
**Pull request by mheinola**
Wrapper for com.vaadin.navigator package.

### CODE REVIEW

File: addon/src/main/scala/vaadin/scala/Navigator.scala [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+
+  /**
+   * @see com.vaadin.navigator.NavigationStateManager
+   */
+  trait NavigationStateManagerMixin extends ScaladinInterfaceMixin {
+    self: com.vaadin.navigator.NavigationStateManager =>
+    override def wrapper = super.wrapper.asInstanceOf[Navigator.NavigationStateManager]
+
+    def getState: String = {
+      wrapper.state.getOrElse("")
```
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala:20]: should be Some(wrapper.state)

File: addon/src/main/scala/vaadin/scala/Navigator.scala [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+  trait ViewMixin extends ScaladinInterfaceMixin {
+    self: com.vaadin.navigator.View =>
+    override def wrapper = super.wrapper.asInstanceOf[Navigator.View]
+
+    def enter(event: com.vaadin.navigator.ViewChangeListener.ViewChangeEvent) {
+      wrapper.enter(Navigator.ViewChangeEvent(
+        WrapperUtil.wrapperFor(event.getNavigator).get,
+        WrapperUtil.wrapperFor(event.getOldView),
+        WrapperUtil.wrapperFor(event.getNewView).get,
+        event.getViewName,
```
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: it seems that viewName can be null so should be Option

File: addon/src/main/scala/vaadin/scala/Navigator.scala [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+      wrapper.showView(WrapperUtil.wrapperFor(view))
+    }
+  }
+
+  trait ViewProviderMixin extends ScaladinInterfaceMixin {
+    self: com.vaadin.navigator.ViewProvider =>
+    override def wrapper = super.wrapper.asInstanceOf[Navigator.ViewProvider]
+
+    override def getViewName(s: String): String = {
+      wrapper.viewName(s).getOrElse(null)
```
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: should be Option(wrapper.viewName(s))

File: addon/src/main/scala/vaadin/scala/Navigator.scala [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+    }
+
+    override def getView(s: String): com.vaadin.navigator.View = {
+      wrapper.view(s).map(_.pView).getOrElse(null)
```
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: should return None if null

File: addon/src/main/scala/vaadin/scala/Navigator.scala [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+
+object Navigator {
+
+  trait NavigationStateManager extends InterfaceWrapper {
+    val pNavigationStateManager = new com.vaadin.navigator.NavigationStateManager with NavigationStateManagerMixin
+    pNavigationStateManager.wrapper = this
+
+    def navigator: Option[Navigator]
+
+    def navigator_=(n: Navigator) {
```
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: javadoc says "This method should only be called by a Navigator." so not needed in scaladin's API

File: addon/src/main/scala/vaadin/scala/Navigator.scala [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+      navigator_=(Some(n))
+    }
+
+    def navigator_=(n: Option[Navigator])
```
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: javadoc says "This method should only be called by a Navigator." so not needed in scaladin's API

File: addon/src/main/scala/vaadin/scala/Navigator.scala [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+
+    def state: Option[String]
+
+    def state_=(s: String) {
```
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: javadoc says "This method should only be called by a Navigator." so not needed in scaladin's API

File: addon/src/main/scala/vaadin/scala/Navigator.scala [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+      state_=(Some(s))
+    }
+
+    def state_=(s: Option[String])
```
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala:104]: javadoc says "This method should only be called by a Navigator." so not needed in scaladin's API

File: addon/src/main/scala/vaadin/scala/Navigator.scala [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+      }
+    }).getOrElse(None)
+
+    def state_=(s: Option[String]) {
+      page.setUriFragment(fragment = "!" + s.getOrElse(""), false)
+    }
+
+  }
+
+  // TODO: extends CustomComponent with ViewDisplay
```
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: What this means?

File: addon/src/main/scala/vaadin/scala/Navigator.scala [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+        case c: Component => {
+          container.removeAllComponents()
+          container.addComponent(c)
+        }
+        case _ => throw new IllegalArgumentException("View is not a component: " + view)
+      })
+    }
+  }
+
+  // TODO: extends CustomComponent with ViewDisplay
```
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: What this means?

File: addon/src/main/scala/vaadin/scala/Navigator.scala [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+  }
+
+  def apply(ui: UI, container: ComponentContainer): Navigator = {
+    new Navigator(ui, new ComponentContainerViewDisplay(container))
+  }
+
+  def apply(ui: UI, container: SingleComponentContainer): Navigator = {
+    new Navigator(ui, new SingleComponentContainerViewDisplay(container))
+  }
+
```
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: provide more apply methods? the same as Vaadin constructors provide

File: addon/src/main/scala/vaadin/scala/Navigator.scala [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+  def addView(viewName: String, viewClass: Class[_ <: Navigator.View]) {
+    addViewProvider(viewName, new Navigator.ClassBasedViewProvider(viewName, viewClass))
+  }
+
+  private def addViewProvider(viewName: String, provider: Navigator.ViewProvider) {
+    removeView(viewName)
+    addProvider(provider)
+  }
+
+  def removeView(viewName: String) {
```
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala:310]: can viewName be null? if so provide also removeView(viewName: Option[String])

File: addon/src/main/scala/vaadin/scala/Navigator.scala [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+        try {
+          Some(viewClass.newInstance())
+        } catch {
+          case e: Exception => throw new RuntimeException(e)
+        }
+      }
+    }.pViewProvider)
+  }
+
+  def errorView(view: Navigator.View) {
```
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: rename to setErrorView

File: addon/src/main/scala/vaadin/scala/Navigator.scala [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+
+  def addProvider(provider: Navigator.ViewProvider) {
+    p.addProvider(provider.pViewProvider)
+  }
+
+  def removeProvider(provider: Navigator.ViewProvider) {
+    p.removeProvider(provider.pViewProvider)
+  }
+
+  def errorView(viewClass: Class[_ <: Navigator.View]) {
```
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: rename to setErrorView

File: addon/src/main/scala/vaadin/scala/Navigator.scala [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+    }
+
+    def showView(v: Option[Navigator.View])
+  }
+
+  trait ViewProvider extends InterfaceWrapper {
+    val pViewProvider = new com.vaadin.navigator.ViewProvider with ViewProviderMixin
+    pViewProvider.wrapper = this
+
+    def viewName(viewAndParameters: String): Option[String]
```
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: getViewName better name? viewAndParameters can be null so provide implemented viewName(viewAndParameters: Option[String]) too

File: addon/src/main/scala/vaadin/scala/Navigator.scala [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+
+    def view(name: String): Option[Navigator.View]
```
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: see previous comment, same apply here

File: addon/src/main/scala/vaadin/scala/Navigator.scala (hunk 12 above)
Comment (mheinola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala:310]: com.vaadin.navigator.Navigator#removeView(String viewName) method only applies to views registered using #addView(String, View) or #addView(String, Class) which both have a null check:
if (viewName == null || view == null) {
            throw new IllegalArgumentException(
But I added def removeView(viewName: Option[String]) anyway. Should I remove it?

File: addon/src/main/scala/vaadin/scala/Navigator.scala (hunk 2 above)
Comment (mheinola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: Done, also for the parameters.

File: addon/src/main/scala/vaadin/scala/Navigator.scala (hunk 5 above)
Comment (mheinola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: stateManager.navigator = this is called on both vaadin and scaladin constructors and also the interface cannot mixin if implementation is missing.

File: addon/src/main/scala/vaadin/scala/Navigator.scala (hunk 6 above)
Comment (mheinola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: stateManager.navigator = this is called on both vaadin and scaladin constructors and also the interface cannot mixin if implementation is missing.

File: addon/src/main/scala/vaadin/scala/Navigator.scala (hunk 11 above)
Comment (mheinola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: Done

File: addon/src/main/scala/vaadin/scala/Navigator.scala (hunk 13 above)
Comment (mheinola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: Renamed this and the rest get/set methods that are not basic getters and/or setters.

File: addon/src/main/scala/vaadin/scala/Navigator.scala (hunk 15 above)
Comment (mheinola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: Done (and for the next comment also)

File: addon/src/main/scala/vaadin/scala/Navigator.scala (hunk 12 above)
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala:310]: You are correct, removeView(viewName: Option[String]) is not needed.

File: addon/src/main/scala/vaadin/scala/Navigator.scala (hunk 1 above)
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala:20]: I was wrong, you code is correct

File: addon/src/main/scala/vaadin/scala/Navigator.scala (hunk 3 above)
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: I was wrong, your code is correct.

File: addon/src/main/scala/vaadin/scala/Navigator.scala (hunk 4 above)
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: I was wrong, your code is correct.

File: addon/src/main/scala/vaadin/scala/Navigator.scala (hunk 5 above)
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: Yes, I think that your implementation is actually good.

File: addon/src/main/scala/vaadin/scala/Navigator.scala (hunk 6 above)
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: Yes, I think that your implementation is actually good.

File: addon/src/main/scala/vaadin/scala/Navigator.scala (hunk 7 above)
Comment (henrikerola) [inline addon/src/main/scala/vaadin/scala/Navigator.scala]: Yes, I think that your implementation is actually good.

**Comment by henrikerola** [comment]
Thanks a lot!

**Comment by mheinola** [comment]
Thank you, I bet that the time it took you to tutor me to get it done you could have done a couple of Navigators yourself :)

**Pull request merged**